lib/srv/discovery/fetchers/aws-sync/kms.go

```diff
@@ -0,0 +1,214 @@
+	// This is a temporary solution until we have a better way to limit the
+	// number of concurrent requests.
+	errGroup.SetLimit(5)
+
+	for _, region := range a.Regions {
+		errGroup.Go(func() error {
+			awsCfg, err := a.AWSConfigProvider.GetConfig(ctx, region, a.getAWSOptions()...)
+			if err != nil {
+				collectKeys(nil, trace.Wrap(err))
+				return nil
+			}
+			client := a.awsClients.getKMSClient(awsCfg)
+			a.collectKMSKeys(ctx, client, collectKeys, region)
+			return nil
+		})
+	}
+
+	err := errGroup.Wait()
+	return keys, trace.NewAggregate(append(errs, err)...)
+}
+
+type kmsKeyCollector func(newKeys []*pb.AWSKMSKeyV1, errs ...error)
+
+func (a *Fetcher) collectKMSKeys(ctx context.Context, client kmsClient, collectKeys kmsKeyCollector, region string) {
+	input := &kms.ListKeysInput{}
+	opt := func(opt *kms.ListKeysPaginatorOptions) { opt.StopOnDuplicateToken = true }
+	pager := kms.NewListKeysPaginator(client, input, opt)
+
+	var keys []*pb.AWSKMSKeyV1
+	var errs []error
+	for pager.HasMorePages() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			collectKeys(nil, trace.Wrap(err))
+			return
+		}
+		for _, key := range page.Keys {
+			key, err := a.fetchKMSKey(ctx, client, key.KeyId, region)
+			if err != nil {
+				collectKeys(nil, err)
+			}
+			if key != nil {
+				keys = append(keys, key)
+			}
+		}
+	}
+	collectKeys(keys, errs...)
+}
+
+func (a *Fetcher) fetchKMSKey(ctx context.Context, client kmsClient, keyID *string, region string) (*pb.AWSKMSKeyV1, error) {
+	input := &kms.DescribeKeyInput{KeyId: keyID}
+	output, err := client.DescribeKey(ctx, input)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to describe KMS key %q", *keyID)
+	}
+	var errs []error
+	result := awsToProtoKMSKey(output, a.AccountID, region)
+	result.Tags, err = getTags(ctx, client, keyID)
+	if err != nil {
+		errs = append(errs, trace.Wrap(err, "cannot fetch tags for KMS key %q", *keyID))
+	}
+	result.Aliases, err = getAliases(ctx, client, keyID)
+	if err != nil {
+		errs = append(errs, trace.Wrap(err, "cannot fetch aliases for KMS key %q", *keyID))
+	}
+	result.PolicyDocument, err = getPolicy(ctx, client, keyID)
+	if err != nil {
+		errs = append(errs, trace.Wrap(err, "cannot fetch policy for KMS key %q", *keyID))
+	}
+	if len(errs) > 0 {
+		return result, trace.NewAggregate(errs...)
+	}
+	return result, nil
+}
+
+func awsToProtoKMSKey(output *kms.DescribeKeyOutput, accountID, region string) *pb.AWSKMSKeyV1 {
+	var multiRegionType pb.MultiRegionKeyType
+	cfg := output.KeyMetadata.MultiRegionConfiguration
+	switch {
+	case cfg == nil:
+		multiRegionType = pb.MultiRegionKeyType_MULTI_REGION_KEY_TYPE_NONE // single region
+	case cfg.MultiRegionKeyType == types.MultiRegionKeyTypePrimary:
+		multiRegionType = pb.MultiRegionKeyType_MULTI_REGION_KEY_TYPE_PRIMARY
+	case cfg.MultiRegionKeyType == types.MultiRegionKeyTypeReplica:
+		multiRegionType = pb.MultiRegionKeyType_MULTI_REGION_KEY_TYPE_REPLICA
+	}
```

Comment:
Do we really need an enum to represent a type? It's probably fine since this is an internal proto message but would there be a downside to simplifying this and just using a plain string?

Reply:
Good idea, done.